src/main/java/com/kovalevskyi/java/deep/model/graph/Neuron.java

```diff
@@ -3,11 +3,9 @@
 import java.util.Map;
 import java.util.Set;
 
-public interface Neuron {
+public abstract class Neuron {
```

Comment:
Converting this to an interface seems unnecessary since all of the methods are abstract anyway. As an aside, if `Neuron` Is part of a public-facing API, it might be best left as an Interface for the sake of the user.

Reply:
yep, you're right! Though, there is a reason for this action, let me share it with you, perhaps you can help me to find the best possible way forward. 

Let start with the example. if someone calls:

```java
neuronA.connect(neuronB, 0.5);
```

what happens next is:
1. forward connection created:
 ```neuronA =====> neuronB```
2. backward connection (with weight is added):
  ```neuronA <==0.5== neuronB```

While doing step1 is simple, doing step2 require to change the internal state of the neuronB. In order to do so there should be a method, however I do not want this method to be public. As a result abstract class allows to create abstract package local method (```addBackwardConnection```). The method is used during the step2.

So as can be seen, the main reason of going from the interface to the abstract class is to have non-public method ```addBackwardConnection``` for in-package use only.

This looks ugly indeed :( any ideas how this can be done better?